Add chapter on NLP #137 (#173)
se-edu/learningresources: contents/naturalLanguage/nl.md

<frontmatter>
  title: Introduction to Natural Language Processing (NLP)
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Natural Language Processing (NLP)
Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)

* [What is NLP](#what-is-nlp)
* [Core problems in NLP](#core-problems-in-nlp)
   * [Syntax](#syntax)
   * [Semantics](#semantics)
   * [Relations](#relations)
* [Applications of NLP](#applications-of-nlp)
* [What's next](#whats-next)
* [References](#references)

> ## What is NLP

Reviewer: I think we're standardising questions to have question marks so it should be `What is NLP?`

Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
 
NLP draws from many other fields of science, from formal linguistics to statistics. 
The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.

> Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes.

Reviewer: Might be good to state what the three themes are, i.e. "most of them can be generalized to three themes: Syntax, Semantics and Relations, which are discussed in the next section"

## Themes in NLP

### Syntax
In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.

> ![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>

Reviewer: Might be good to increase the image size and add a subheader for an image using the 'pic' syntax below.

```xml
<pic src="partsOfSpeech.png" alt="partsOfSpeech" width="70%">

  <sub> _Figure 1. Parts of Speech_ </sub>

</pic>
```

It would also be more aesthetic to centre the image.

For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.

![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>

> From the parts-of-speech tags, a tree-structured representation of grammar can be produced.

Reviewer: Likewise, can add a subheader here

This tree is the result of a _constituency parse_, a task of recognizing a sentence and assigning a syntactic structure to it.
From the tree we can see that the complex phrase 'word-level parts-of-speech' is simply a _noun phrase_.
Incidentally, this is how Microsoft Word checks for grammatical errors<sup>[6](#footnote6)</sup>.
Any tree that cannot be parsed may be grammatically incorrect or difficult to understand.

### Semantics
With the knowledge of a sentence and its structure, the next step is to understand the meaning it conveys.
To this end, the representation of the meaning of a sentence should be able to link language to concepts<sup>[7](#footnote7)</sup>.

Take for example the sentence: 

   > I wrote an article!

In order to understand the meaning of the sentence, a few questions need to be answered:

* Who is 'I'?
* What is 'an article'? 
* How are the two subjects related?

An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.

![Relation tuple](openIE.png "Constituency Parse") <br/>

> From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that 'I' have written".

Reviewer: "What are the articles that 'I' have written?"

There are attempts to create semantic _ontologies_ for natural language, groupings of words by their meanings.
**WordNet** is one such attempt, a handcrafted database that classifies words by concepts that they express<sup>[9](#footnote9)</sup>.

![WordNet](wordNet.png "WordNet") <br/>

More recently, techniques like **word2vec**<sup>[10](#footnote10)</sup> are able to model words and phrases as vectors of real numbers. 
This ability to quantify and categorize semantic similarity allows computers to infer beyond just synonyms.

![word2vec](word2vec.png "Word2Vec") <br/>

A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.

Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.

### Applications of NLP

#### Input methods editors, e.g. Chinese, Japanese and Korean

> Modern input method editors(IMEs) do more than simply translating input to output in the respective language.

Reviewer: Might be good to elaborate on what Method Input Editors are, since it's introduced here for the first time

#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 

#### Machine translation, e.g. Google Translate
**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
This family of techniques uses special neural network architectures to model sentences.  

#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 
For example, the two possible input from users regarding a replacement for a digital banking token: 

> * I want to replace my token.
> * My old token is broken. 

can be mapped back to the same intent to which a predefined response can be given. 

![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>

## What's next

> 1. Get an intuition for the English language

Reviewer: Might be good to bold each of the points to make it stand out more

    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
    This is invaluable especially for those who grew up with English as a first language.

1. Visually explore the themes present in NLP
    

>     Standford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.

Reviewer: Stanford's

machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
 
NLP draws from many other fields of science, from formal linguistics to statistics. 
The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.

Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes.

## Themes in NLP

### Syntax
In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.

![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.

> ![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>

Reviewer 2: Could the blue borders be removed, or surround the entire image?

From the parts-of-speech tags, a tree-structured representation of grammar can be produced.

> This tree is the result of a _constituency parse_, an task of recognizing a sentence and assigning a syntactic structure to it.

Reviewer 2: `an task` -> `a task`

From the tree we can see that the complex phrase 'word-level parts-of-speech' is simply a _noun phrase_.
Incidentally, this is how Microsoft Word checks for grammatical errors<sup>[6](#footnote6)</sup>.
Any tree that cannot be parsed may be grammatically incorrect or difficult to understand.

### Semantics
With the knowledge of a sentence and its structure, the next step is to understand the meaning it conveys.
To this end, the representation of the meaning of a sentence should be able to link language to concepts<sup>[7](#footnote7)</sup>.

Take for example the sentence: 

   > I wrote an article!

In order to understand the meaning of the sentence, a few questions need to be answered:

* Who is 'I'?
* What is 'an article'? 
* How are the two subjects related?

An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.

![Relation tuple](openIE.png "Constituency Parse") <br/>

> From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that I have written".

Reviewer 2: is the structure in `What the articles that I have written` intentional?

Author: Added quotes around the 'I' similar to the previous examples for emphasis.

This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>

### Relations 

> Relations is crucial to the understanding of natural languages.

Reviewer 2: `are`

![WordNet](wordNet.png "WordNet") <br/>

More recently, techniques like **word2vec**<sup>[10](#footnote10)</sup> are able to model words and phrases as vectors of real numbers. 
This ability to quantify and categorize semantic similarity allows computers to infer beyond just synonyms.

![word2vec](word2vec.png "Word2Vec") <br/>

A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.

Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.

### Applications of NLP

#### Input methods editors, e.g. Chinese, Japanese and Korean
Modern input method editors(IMEs) do more than simply translating input to output in the respective language.
By tackling the syntactic and relational elements of natural language, IMEs provide autocomplete functionality by suggesting words that are related or occur together frequently. 

> For example, typing in "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.

Reviewer 2: ~in~

#### Indexing and information retrieval, e.g. Google's page rank and normalization of search terms
Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
This is achieved with techniques similar to **word2vec**. 

#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 

#### Machine translation, e.g. Google Translate
**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
This family of techniques uses special neural network architectures to model sentences.  

#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 

> For example, the two possible input from users regarding a replacement of digital banking token: 

Reviewer 2: `of a`

> * I want to replace my token.
> * My old token is broken. 

![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>

> can be mapped back to the same intent to which a predefined response can be given. 

Reviewer 2: Could move this above the second image

## What's next
1. Get an intuition for the English language

>     The Language Instinct by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 

Reviewer 2: Missing link here

<frontmatter>
  title: Introduction to Natural Language Processing (NLP)
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Natural Language Processing (NLP)
Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)

> * [What is NLP?](#what-is-nlp)

Reviewer 3: Can wrap this TOC in a 
```
<box id="article-toc">`
... 
</box>
```
, this will ensure that the TOC is responsive with the screen re-sizing

> * [Core problems in NLP](#core-problems-in-nlp)

Reviewer 3: The section Core problems in NLP does not match with the Themes in NLP header

   * [Syntax](#syntax)
   * [Semantics](#semantics)
   * [Relations](#relations)
* [Applications of NLP](#applications-of-nlp)
* [What's next](#whats-next)
* [References](#references)

## What is NLP?

> Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.

Reviewer 4: ```suggestion
Natural Language Processing (NLP) is the set of methods that makes human language accessible to computers.
```

Author: "for making" has the implication that the methods were tailored for the said purpose.
"that makes" loses a bit of that meaning IMO.

Reviewer 4: Ah I got your point. I just feel the sentence is a bit weirdly framed.

Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
 
NLP draws from many other fields of science, from formal linguistics to statistics. 
The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.

> Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: syntax, semantics and relations.

Reviewer 3: perhaps can italicise the syntax, semantics, relations and allow them to be clickable that navigates to their corresponding sections?

## Themes in NLP

### Syntax
In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.

<p align="center">
    <pic src="partsOfSpeech.png" alt="partsOfSpeech" width="90%">
      <sub><i>Figure 1. Parts of Speech</i></sub>
    </pic>
</p>

> For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.

Reviewer 3: in the preview, the footnote 5 seems to not work
<img width="197" alt="Screen Shot 2020-03-27 at 12 44 54 PM" src="https://user-images.githubusercontent.com/32864116/77722458-c38caa80-7028-11ea-8049-e16e8fa3d962.png">

<p align="center">
    <pic src="wordNet.png" alt="WordNet results for write" width="90%">
      <sub><i>Figure 4. WordNet results for the word "write"</i></sub>
    </pic>
</p>

More recently, techniques like **word2vec**<sup>[10](#footnote10)</sup> are able to model words and phrases as vectors of real numbers. 
This ability to quantify and categorize semantic similarity allows computers to infer beyond just synonyms.

<p align="center">
    <pic src="word2vec.png" alt="word2vec visualization for antonyms" width="90%">
      <sub><i>Figure 5. word2vec visualization for antonyms</i></sub>
    </pic>
</p>

A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.

Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.

> ### Applications of NLP

Reviewer 4: All headings in this section are quite lengthy. Can you prune them?

For example, typing "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.

#### Indexing and information retrieval, e.g. Google's page rank and normalization of search terms
Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
This is achieved with techniques similar to **word2vec**. 

#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 

#### Machine translation, e.g. Google Translate
**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
This family of techniques uses special neural network architectures to model sentences.  

#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep

> Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 

Reviewer 3: mistake of putting 14 instead of 15

For example, the two possible input from users regarding a replacement for a digital banking token: 

> * I want to replace my token.
> * My old token is broken. 

can be mapped back to the same intent to which a predefined response can be given. 

<p align="center">
    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
    </pic>
</p>

> ## What's next

Reviewer 4: ```suggestion
## What's Next?
```

<frontmatter>
  title: Introduction to Natural Language Processing (NLP)
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Natural Language Processing (NLP)
Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)

<box id="article-toc">
* [What is NLP?](#what-is-nlp)
* [Themes in NLP](#themes-in-nlp)
   * [Syntax](#syntax)
   * [Semantics](#semantics)
   * [Relations](#relations)
* [Applications of NLP](#applications-of-nlp)
* [What's Next](#whats-next)
* [References](#references)
</box>

## What is NLP?
Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;

> machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 

Reviewer 5: These references feel a bit random. Usually, the reference should point to the source of the fact given in the sentence. If they are simply examples, best to give them specifically as examples.

 
NLP draws from many other fields of science, from formal linguistics to statistics. 
The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.

Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: [_syntax_](#syntax), [_semantics_](#semantics) and [_relations_](#relations).

## Themes in NLP

### Syntax

> In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.

Reviewer 5: Abrupt usage of ML here. The relation is not explained above.

Author: Rephrased to match the definition given in the intro.

In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.

> <p align="center">

Reviewer 3: can use the md syntax `<center></center>` to center the images

    <pic src="partsOfSpeech.png" alt="partsOfSpeech" width="90%">
      <sub><i>Figure 1. Parts of Speech</i></sub>
    </pic>
</p>

> For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[6](#footnote5)</sup>.

Reviewer 3: should this be `[5]` instead?

<p align="center">
    <pic src="constituencyParse.png" alt="Constituency Parse" width="90%">
      <sub><i>Figure 2. Constituency Parse</i></sub>
    </pic>
</p>

From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
This tree is the result of a _constituency parse_, a task of recognizing a sentence and assigning a syntactic structure to it.
From the tree we can see that the complex phrase 'word-level parts-of-speech' is simply a _noun phrase_.
Incidentally, this is how Microsoft Word checks for grammatical errors<sup>[6](#footnote6)</sup>.

> Any tree that cannot be parsed may be grammatically incorrect or difficult to understand.

Reviewer 5: Should have used an unrelated sentence in the example so as not to add unnecessary complications (never go meta with examples as it always make the reader's job harder). Was this example taken from elsewhere? if so, cite the source. If you designed the example, consider using a different sentence.

Author: Changed it to a simpler sentence.

### Semantics
With the knowledge of a sentence and its structure, the next step is to understand the meaning it conveys.
To this end, the representation of the meaning of a sentence should be able to link language to concepts<sup>[7](#footnote7)</sup>.

Take for example the sentence: 

   > I wrote an article!

In order to understand the meaning of the sentence, a few questions need to be answered:

* Who is 'I'?
* What is 'an article'? 
* How are the two subjects related?

An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.

<p align="center">
    <pic src="openIE.png" alt="Relation tuple" width="90%">
      <sub><i>Figure 3. Relation tuple</i></sub>
    </pic>
</p>

From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What are the articles that 'I' have written".

> This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>

Reviewer 5: _Watson uses this technique to defeat human player_ seems to convey a misleading message. Isn't it _Watson, who is able to defeat human players, uses this technique_? Even then, I'm not sure there is any direct link between the technique and defeating the human player.

Author: Changed the example to some other downstream application.

### Relations 
Relations are crucial to the understanding of natural languages.
Consider the example:

   > Geoffrey bought Mary a ring. They have been dating for months.

What is the event that prompted Geoffrey to buy a ring? 
The word _proposal_, while not stated in the example, implicitly links the two events together.
While humans are able to draw the connections instinctually, it is not easy to formalize this process computationally.

> There are attempts to create semantic _ontologies_ for natural language, groupings of words by their meanings.

Reviewer 5: Unexplained term: ontology

Reviewer 5: Explanation of new terms can be given as tooltips, BTW.

Author: Added a tooltip.

<p align="center">
    <pic src="wordNet.png" alt="WordNet results for write" width="90%">
      <sub><i>Figure 4. WordNet results for the word "write"</i></sub>
    </pic>
</p>

More recently, techniques like **word2vec**<sup>[10](#footnote10)</sup> are able to model words and phrases as vectors of real numbers. 
This ability to quantify and categorize semantic similarity allows computers to infer beyond just synonyms.

<p align="center">
    <pic src="word2vec.png" alt="word2vec visualization for antonyms" width="90%">
      <sub><i>Figure 5. word2vec visualization for antonyms</i></sub>
    </pic>
</p>

A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.

Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.

> ### Applications of NLP

Reviewer 5: This heading should be level 2?

By tackling the syntactic and relational elements of natural language, IMEs provide autocomplete functionality by suggesting words that are related or occur together frequently. 

For example, typing "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.

#### Indexing and information retrieval, e.g. Google's page rank and normalization of search terms
Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
This is achieved with techniques similar to **word2vec**. 

#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 

#### Machine translation, e.g. Google Translate
**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 

> This family of techniques uses special neural network architectures to model sentences.  

Reviewer 5: _an_ algorithm?

> This family of techniques uses special neural network architectures to model sentences.  

Reviewer 5: Unexplained term: neural network

Author: Added a tooltip.

#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[15](#footnote15)</sup>. 
For example, the two possible input from users regarding a replacement for a digital banking token: 

> * I want to replace my token.
> * My old token is broken. 

can be mapped back to the same intent to which a predefined response can be given. 

<p align="center">
    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
    </pic>
</p>

## What's Next

> 1. **Get an intuition for the English language**

Reviewer 5: So, NLP is tied to English?

Author: Rephrased the point to include linguistics in general.

    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 

>     This is invaluable especially for those who grew up with English as a first language.

Reviewer 5: The reason for this is not obvious.

Author: Rephrased the point to be multilingual.

1. **Visually explore the themes present in NLP**
    
    Stanford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.
    Explore how sentences are annotated with *parts-of-speech* tags and see the output of a constituency parse.
    
1. **Experiment with NLP libraries**

    The aforementioned [CoreNLP](https://stanfordnlp.github.io/CoreNLP/) library is a good starting point for developers comfortable in Java.
    For a simple and productive experience, [spaCy](https://spacy.io/) is a Python library suitable for experimentation and rapid prototyping.
    
1. **Read research papers to gain a deeper understanding of the techniques employed**

>     Stay up to date with research through portals like [arxiv sanity preserver](https://www.arxiv-sanity.com/search?q=nlp).

Reviewer 5: I'm not sure we should pretend to guide the reader to _self-learn_  NLP. Is that even practical? Perhaps just point the reader to a good online NLP course?

Author: Included a reference to CS224n by Stanford.

    
    
## References

<a name="footnote1">[1]</a>: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6562150/ <br/>
<a name="footnote2">[2]</a>: https://azure.microsoft.com/en-us/services/bot-service/ <br/>

> <a name="footnote3">[3]</a>: https://github.com/google/seq2seq <br/>

Reviewer 5: Same as 14?

Author: Removed.

By tackling the syntactic and relational elements of natural language, IMEs provide autocomplete functionality by suggesting words that are related or occur together frequently. 

For example, typing "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.

#### Indexing and information retrieval, e.g. Google's page rank and normalization of search terms
Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
This is achieved with techniques similar to **word2vec**. 

#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 

#### Machine translation, e.g. Google Translate
**seq2seq**<sup>[3](#footnote3)</sup>, a algorithm developed by Google, transforms a sequence to another. 

> This family of techniques uses special <tooltip content="artifical neural networks a computer system made up of a number of simple, highly connected processing elements which translate information to some form of desired output"><i>artificial neural network</i></tooltip>   architectures to model sentences.  

Reviewer 5: >artifical neural networks a computer system made up ...

Something missing in that phrase.

#### Automated customer support, e.g. ChatBot, NanoRep
Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote14)</sup>. 
For example, the two possible input from users regarding a replacement for a digital banking token: 

> * I want to replace my token.
> * My old token is broken. 

can be mapped back to the same intent to which a predefined response can be given. 

<p align="center">
    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
    </pic>
</p>

## What's Next
1. **Get an intuition for linguistics**

    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
    Learn about how to deconstruct language into formal structures for critical analysis through engaging examples.
    
1. **Visually explore the themes present in NLP**
    

>     Stanford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.

Reviewer 5: I couldn't get this to run though

Author: It can be a bit sluggish to load at times depending on the load of the server.
